java/src/jmri/jmrix/loconet/messageinterp/LocoNetMessageInterpret.java

```diff
@@ -2323,6 +2343,44 @@ private static String interpretOpcMultiSenseTranspPresence(LocoNetMessage l, Str
         }
     }
 
+    private static String interpretOpcMultiSenseLong(LocoNetMessage l, String reporterPrefix) {
+        if (l.getElement(1) == 0x09){  // Only process 0xE0 0x09 messages
+            // Transponding Event
+            // get system and user names
+            String reporterSystemName;
+            String reporterUserName;
+
+            int type = l.getElement(2) & LnConstants.OPC_MULTI_SENSE_MSG;
+
+            reporterSystemName = reporterPrefix
+                    + ((l.getElement(2) & 0x1F) * 128 + l.getElement(3) + 1);
+
+            Reporter reporter = InstanceManager.getDefault(ReporterManager.class).getReporter(reporterSystemName);
+            reporterUserName = "";
+            if (reporter != null) {
+                String uname = reporter.getUserName();
+                if ((uname != null) && (!uname.isEmpty())) {
+                    reporterUserName = uname;
+                }
+            }
+
+            String locoAddr = convertToMixed(l.getElement(5), l.getElement(4));
+            String transpActivity = (type == LnConstants.OPC_MULTI_SENSE_ABSENT)
+                    ? Bundle.getMessage("LN_MSG_OPC_MULTI_SENSE_TRANSP_HELPER_IS_ABSENT")
+                    : Bundle.getMessage("LN_MSG_OPC_MULTI_SENSE_TRANSP_HELPER_IS_PRESENT");
```

Comment:
The two valid bits of type (i.e. OPC_MULTI_SENSE_MSG bits of byte 2, per line 2353) have four combinations.  Only one of those four is checked here.  Do you have a guarantee that _all three_ other bit encodings _always_ mean the "Present" state?
It may be better to code for "known" encodings of those bits, and for any encoding that is not known, report "Unknown".

Reply:
RE: Given that there are "undecoded" bits in byte 6, and nothing done with byte 7, 

Still to be verified, but I believe the rest of byte 2 (possibly 0x40 or 0x60), byte 6 (besides bit 0x40) plus byte 7 will be used to indicate, specify type (speed vs QoS) and store RailCom Application Dynamic (App:Dyn) information as defined in the NMRA standard and indicated in the Digikeijs manual as included in this message.  

Digikeijs provides the functionality to include or not include App:Dyn information in Multi Sense Long messages. 

See page 24 of the [DR5088RC manual](https://support.digikeijs.com/hc/en-us/articles/360001793757-DR5088RC-Manuals-Downloads)

I have enabled and disabled the features on this page (bullets 3 to 8) and no additional information is included in Multi Sense Long that I have detected.    

Unfortunately certain RailCom features certain features are optional and vendor specific.  And it seems that my ESU decoders to not support RailCom App:Dyn features (speed or QoS).  These feature is not mentioned in ESU manuals or any product descriptions and this seems to be confirmed at this [OpenDCC web page](https://www.opendcc.de/info/railcom/railcom_decoder_overview.html ).  So I’ve ordered some Zimo decoders that appear to support the feature - but I don’t expect these to arrive for 30 to 60 days.

Reply:
RE: The two valid bits of type (i.e. OPC_MULTI_SENSE_MSG bits of byte 2, per line 2353) have four combinations. 

I debated the same before I submitted.  Based on what I've seen on byte 2 with Multi Sense Long (OPC 0xE0), and when compared to the exact same on Multi Sense Standard (OPC 0xD0) we know 2 of the 4 values.

- 0x00 = Absent
- 0x20 = Present

I expect there is a potential value 0x40 and/or 0x60 associated with App:Dyn or something else. And if so, that App:Dyn would likely indicate a train "Present" on the track/block.

Reply:
I’ve been examining packet values to decipher as this OpCode as it’s unpublished; but when I compare Multi Sense Standard (MSS = 0xDO) to Multi Sense Long (MSL = 0xE0) many of the values are identical - with my current combo of the DR5088RC and ESU decoders.  

The DR5088RC has the functionality to send either MSS, or MSL or both.; so with the “both” settings, it’s easy to compare these OpCodes side by side to identify what is the same.  I created several tests including:  

1. With the maximum responder  address(2048)
2. Maximum DCC address 10239. 
3. Added and removed the loco.  
4. Changed the direction of the loco.  

Here are example tests and MSS and MSL side by side.  

**Max Configurable Responder Address = 2049**
_Place on tracks facing west_ 
[D0 2F 7F 00 0A 75]  Transponder address 10 (short) present at LR2048 () (BXP88 Board ID 256 section 8 or the BXPA1 Board ID 2,048 section).
[E0 09 2F 7F 00 0A 40 00 0C]  Transponder address 10 (short) facing West present at LR2048 ().

_Remove from tracks_ 
[D0 0F 7F 00 0A 55]  Transponder address 10 (short) absent at LR2048 () (BXP88 Board ID 256 section 8 or the BXPA1 Board ID 2,048 section).
[E0 09 0F 7F 00 0A 40 00 2C]  Transponder address 10 (short) facing West absent at LR2048 ().

_Place on tracks facing east_
[D0 2F 7F 00 0A 75]  Transponder address 10 (short) present at LR2048 () (BXP88 Board ID 256 section 8 or the BXPA1 Board ID 2,048 section).
[E0 09 2F 7F 00 0A 00 00 4C]  Transponder address 10 (short) facing East present at LR2048 ().

_Remove from tracks_
[D0 0F 7F 00 0A 55]  Transponder address 10 (short) absent at LR2048 () (BXP88 Board ID 256 section 8 or the BXPA1 Board ID 2,048 section).
[E0 09 0F 7F 00 0A 00 00 6C]  Transponder address 10 (short) facing East absent at LR2048 ().

**Max Configurable DCC address = 10239**
_Place on tracks facing East_
[B2 13 50 0E]  Sensor LS39 () is High.  (BDL16 # 3, DS7; DS54/DS64/SE8c # 5, AuxD/A4/DS07).
[D0 20 26 4F 7F 19]  Transponder address 10239 present at LR39 () (BDL16x Board ID 3 RX4 zone D or BXP88 Board ID 5 section 7 or the BXPA1 Board ID 39 section).
[E0 09 20 26 4F 7F 00 00 20]  Transponder address 10239 facing East present at LR39 ().

_Remove from tracks_
[B2 13 40 1E]  Sensor LS39 () is Low.  (BDL16 # 3, DS7; DS54/DS64/SE8c # 5, AuxD/A4/DS07).
[D0 00 26 4F 7F 39]  Transponder address 10239 absent at LR39 () (BDL16x Board ID 3 RX4 zone D or BXP88 Board ID 5 section 7 or the BXPA1 Board ID 39 section).
[E0 09 00 26 4F 7F 00 00 00]  Transponder address 10239 facing East absent at LR39 ().

So from what I’ve seen so far, in MSL (0xE0) bytes 2,3,4,5 are identical to MSS (0xD0) bytes 1,2,3,4.  The only difference byte 6 and its 2nd bit (0x40) indicated direction = Confirmed within Digikeijs GUI direction indicator.  So as a foundation, I started with an understanding that MSL is very similar, and an extension to MSS.

Reply:
From the perspective of decoding a message that is not quite entirely understood, it may be better to place a strict requirement on all "type" bits in byte 2, all bits of byte 6, and all bits of byte 7, in order to interpret the message as you currently do.  Then, as further information comes to light on the LocoNet messages for QoS and Speed (and any other details!), update the code for an appropriate sub-branch and interpretation may be added by comparing to different bit values.

This gives the benefit of decoding messages that fit a known pattern now, while returning some other message for non-matching patterns.  And it helps to preserve the "intent" and "full scope" of whatever "real" specification is in place for the 0xE0 0x09 message type.

For the case where the message is of the 0xE0 0x09 type, but where the message does not match a known pattern, the following sort of return message gives the Monitor message some indication of the message type, but also gives a hex dump of the message.  This can help developers identify differences in the pattern, and with some understanding of the hardware and/or situation which generated the message, help in understanding correct interpretation and generation of new sub-patterns.
`return Bundle.getMessage("LN_MSG_LONG_ACK_OPC_IMM_UNKNOWN", Bundle.getMessage("LN_MSG_HEXADECIMAL_REPRESENTATION",                                     StringUtil.twoHexFromInt(ack1)))+                                 Bundle.getMessage("LN_MONITOR_MESSAGE_RAW_HEX_INFO", l.toString());`

Reply:
Agreed with this logic and will revise as such - I'll add more strict requirements on bytes 2, 6, and 7.  It might take me a few days.  

I also ordered a different Zimo decoder (besides the one mentioned above) so I might have my hands on other App:Dyn RailCom packets within the week.  

I am currently catching and throwing an unhandled monitor on packets with the OpCode 0xE0, but a length (byte 1) other than 0x09. But before I was catching this and throwing my custom `Bundle.getMessage("LN_MSG_OPC_MULTI_SENSE_LONG_UNKNOWN_MESSAGE");`, I was using code found in many other functions that simply call `return "";`.  Is there a situation across LocoNetMessageInterpret.java where we might not catch and report unknown patters very well?  Maybe a separate discussion topic/thread, but unknown packet patterns was something I was planning to look into.  

Thanks for your help.  I'm not in programming by trade, so hacking and learning my way through this.

Reply:
> I am currently catching and throwing an unhandled monitor on packets with the OpCode 0xE0, but a length (byte 1) other than 0x09. 

I am not convinced that throwing an exception is good, since this tool does not do anything other than generate messages for display.  If you throw an exception, that will just make the invoking code have to deal with an exception.

On the other hand, if you find a troublesome condition, simply let the defaulting mechanism kick in (via `return "";`).  This will genearate a message like:
```
[88 77]  Unable to parse LocoNet message.
contents: 88 77
```

Or create your own customized "unable to parse" message return value, like the code snippet I included previously.  I use this format in those cases where the message _should_ be of a particular type but I see something that doesn't match the known pattern.  And I tend to _specifically_ use the customized "unable to parse" message when I think/know that there are other similar contents for which I do not know/understand the encoding.  

I believe that this message is a good example of that last case: how are QoS and Speed encoded in `OPC_MULTI_SENSE_L`?  You do not really know now, so defer that for now via a customized "unable to parse" message.  You know it ought to be OPC_MULTI_SENSE_L, but not exactly how to interpret the rest, so put that in the message.

>  I'm not in programming by trade

Don't let that stop you!  You are not the only non-programmer, who has contributed to the code-base.  I am one.  And I would guess that more than 50% of the code contributors have little/no formal programming training...

Reply:
Still seeing cases where `type = 0x10` or `type = 0x30` (incorrectly?) defaulting to `LN_MSG_OPC_MULTI_SENSE_TRANSP_HELPER_IS_PRESENT`.
Still seeing cases where  `(l.getElement(6) & 0x3F) != 0` allowing message interpretation as `LN_MSG_OPC_MULTI_SENSE_LONG_TRANSP_REPORT`.

Reply:
Code currently `type` reasonably for all cases.

Still seeing cases where (l.getElement(6) & 0x3F) != 0 allowing message interpretation as LN_MSG_OPC_MULTI_SENSE_LONG_TRANSP_REPORT.  This still seems like a bad idea!

Reply:
Addressed with the latest code. For Absent (0x00) or present (0x20) type reports `if ((l.getElement(6) & 0x3F) != 0 || l.getElement(7) != 0)` I now use LN_MSG_OPC_MULTI_SENSE_LONG_UNKNOWN_MESSAGE.  

I also added test cases for this unexpected case and additional test cases with minimum and maximum values.